cranelift-codegen/src/ir/constant.rs

```diff
@@ -0,0 +1,203 @@
+//! Constants
+//!
+//! The constant pool defined here allows cranelift to avoid emitting the same constant multiple
+//! times. As constants are inserted in the pool, a handle is returned; the handle is a cranelift
+//! Entity. Inserting the same data multiple times will always return the same handle. Future work
+//! could include: ensuring alignment of constants within the pool, bucketing constants by size.
+
+use crate::ir::Constant;
+use cranelift_entity::EntityRef;
+use std::collections::{BTreeMap, HashMap};
+use std::vec::Vec;
+
+/// This type describes the actual constant data
+pub type ConstantData = Vec<u8>;
+
+/// This type describes an offset in bytes within a constant pool
+pub type ConstantOffset = u32;
+
+/// Inner type for storing data and offset together in the constant pool. The offset is optional
+/// because it must be set relative to the function code size (i.e. constants are emitted after the
+/// function body); because the function is not yet compiled when constants are inserted,
+/// [set_offset](ir::ConstantPool::set_offset) must be called once a constant's offset from the
+/// beginning of the function is known (see [relaxation.rs](binemit::relaxation)).
+#[derive(Clone)]
+struct ConstantPoolEntry {
+    data: ConstantData,
+    offset: Option<ConstantOffset>,
+}
+
+impl ConstantPoolEntry {
+    fn new(data: ConstantData) -> Self {
+        ConstantPoolEntry { data, offset: None }
+    }
+}
+
+/// Maintains the mapping between a constant handle (i.e. [Constant](ir::entities::Constant)) and
+/// its constant data (i.e. [ConstantData](ir::constant::ConstantData))
+#[derive(Clone)]
+pub struct ConstantPool {
+    // this mapping maintains the insertion order as long as Constants are created with sequentially increasing integers
```

Comment:
When a comment forms a substantial sentence, please capitalize it and end it with a period.